src/pathways/escher.service.ts

```diff
@@ -1,100 +1,135 @@
-						currentThis.map.new_reaction_from_scratch(
-							this.reactions[i].id,
-							this.startCoordinates,
-							this.direction
-						);
-					} else {
-						for (let k in currentThis.map.nodes) {
-							if (currentThis.map.nodes[k].name === this.metabolitesNames[i]) {
-								currentThis.map.new_reaction_for_metabolite(
-									this.reactions[i].id,
-									k,
-									this.direction);
-								break;
-							}
-						}
-					}
-				}
-				for (let n in currentThis.map.nodes) {
-					currentThis.map.nodes[n].node_is_primary =
-						(this.metabolitesNames.indexOf(currentThis.map.nodes[n].name) != -1);
-				}
-				currentThis.map.draw_everything();
-				currentThis.map.zoom_extent_canvas();
-				currentThis.map.select_none();
-			}
-		}
-	};
+    /** Flips the reaction in the other direction by changing
+     * the sign of metabolites' the stoichiometric number
+     */
+    reverseReaction(reaction) {
+        return fp.assign(reaction, {
+            metabolites: _.mapValues(reaction.metabolites, v => -v)
+        });
+    }
 
-	alignReactions(data) {
-		let allReactions = {};
-		for (let i in data.model.reactions) {
-			allReactions[data.model.reactions[i].id] = data.model.reactions[i];
-		}
-		let sortedReactions = data.reactions;
-		for (let i in sortedReactions) {
-			let reaction = allReactions[sortedReactions[i].id];
-			if (reaction.metabolites[data.primary_nodes[i].id] > 0) {
-				for (let k in reaction.metabolites) {
-					reaction.metabolites[k] *= -1;
-				}
-			}
-		}
-		return data.model;
-	}
+    // TODO this violates the SRP -> should be two separate functions.
+    /** Sorts the model's reaction and flips them if required */
+    alignReactions(data) {
+        // The keyBy + at function could be combined into a utility function
+        const allReactions = _.keyBy(data.model.reactions, 'id');
+        const reactionWithPrimaryNode = _.zip(
+            _.at(allReactions, _.map(data.reactions, 'id')),
+            _.map(data.primary_nodes, 'id')
+        );
+        const reactions = _.map(reactionWithPrimaryNode, ([reaction, primary_node]) => {
+            if (reaction.metabolites[primary_node] > 0) {
+                return this.reverseReaction(reaction);
+            } else {
+                return reaction;
+            }
+        });
+        return fp.assign(data.model, {reactions: reactions});
+    }
 
-	buildMap(data, controller_id) {
-		this.metabolitesNames = [];
-		data.primary_nodes.forEach(
-			(value) => this.metabolitesNames.push(value.name)
-		);
-		if (this.reactions == data.reactions) {
-			return;
-		}
-		this.reactions = data.reactions;
-		escher.Builder(
-			null,
-			this.alignReactions(data),
-			null,
-			d3.select('#' + controller_id),
-			this.options
-		);
-	}
+    buildMap(data, controller_id) {
+        this.metabolitesNames = data.primary_nodes
+            .map(node => node.name);
+        // Looks like an early return. When does it happen?
```

Comment:
Sounds like a good comment, since the map doesn't affect either `this.reactions` or `data.reactions` I think this early return should be moved to the beginning of the function and formatted into one line.